### PR TITLE
istiod: server startup more refactor

### PR DESCRIPTION
**Pull request by ramaraochavali**
Continuing on #23492 and #23684 this PR refactors it more and does the following
- Move initHandlers, Debug server out of `initDiscoveryService` as they do not belong there.
- Fixes close functionality (remove forceStop flag and always stop gracefully (with less duration in tests), closes DNS server, secure gRPC server and handles gRPC shutdown correctly
- Removes the dependency of webhook on server by refactoing readiness probe into generic readiness probes
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -389,42 +399,15 @@ func (s *Server) httpServerReadyHandler(w http.ResponseWriter, _ *http.Request)
...
-
-	// Implement EnvoyXdsServer grace shutdown
-	s.addStartFunc(func(stop <-chan struct{}) error {
-		s.EnvoyXdsServer.Start(stop)
-		return nil
-	})
-
+// initDebugServer initializes debug end points.
+func (s *Server) initDebugServer(args *PilotArgs, wh *inject.Webhook) error {
+	s.debugMux.HandleFunc("/ready", s.httpServerReadyHandler)
```
Comment (costinm) [inline pilot/pkg/bootstrap/server.go:404]: Technically '/ready' is not part of the debug server - but readiness. 

I remember in a distant past the 'debug' interfaces had a flag that allowed disabling them - maybe we should bring it back. 

I think the /ready may share the same port with monitoring - both need to be enabled.

Comment (ramaraochavali) [reply pilot/pkg/bootstrap/server.go]: Changed it to IstiodAdminServer combining monitoring, debug and readiness as is being done today

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -102,6 +102,9 @@ func init() {
 // startFunc defines a function that will be used to start one or more components of the Pilot discovery service.
 type startFunc func(stop <-chan struct{}) error
 
+// readinessProbe defines a function that will be used indicate whether a server is ready.
+type readinessProbe func() int
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/bootstrap/server.go:106]: Yeah. I also got the same impression when I wrote it. Status code is needed to respond back. I will indicate that this is status code

**Comment by howardjohn** [comment]
You could just return 503 if any function returns false. Not a big issue
though

**Comment by ramaraochavali** [comment]
> You could just return 503 if any function returns false. Not a big issue
though

that makes sense. https://github.com/istio/istio/pull/24005 PTAL

**Pull request merged**